ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Forge.hs

```diff
@@ -31,19 +47,152 @@ import           Ouroboros.Consensus.Shelley.Ledger.Config
+-- | Get the KES key from the node state, evolve if its KES period doesn't
+-- match the given one.
+evolveKESKeyIfNecessary
+  :: forall m c. (MonadRandom m, TPraosCrypto c)
+  => Update m (TPraosForgeState c)
+  -> SL.KESPeriod -- ^ Relative KES period (to the start period of the OCert)
+  -> m (SignKeyKES (KES c))
+evolveKESKeyIfNecessary updateForgeState (SL.KESPeriod kesPeriod) = do
+    getOudatedKeyOrCurrentKey >>= \case
+      Right currentKey -> return currentKey
+      Left outdatedKey -> do
+        let newKey@(HotKey _ key) = evolveKey outdatedKey
+        saveNewKey newKey
+        return key
+  where
+    -- | Return either (@Left@) an outdated key with its current period (setting
+    -- the node state to 'TPraosKeyEvolving') or (@Right@) a key that's
+    -- up-to-date w.r.t. the current KES period (leaving the node state to
+    -- 'TPraosKeyAvailable').
+    getOudatedKeyOrCurrentKey
+      :: m (Either (HotKey c) (SignKeyKES (KES c)))
+    getOudatedKeyOrCurrentKey = runUpdate updateForgeState $ \case
+      TPraosKeyEvolving ->
+        -- Another thread is currently evolving the key; wait
+        Nothing
+      TPraosKeyAvailable hk@(HotKey kesPeriodOfKey key)
+        | kesPeriodOfKey < kesPeriod
+          -- Must evolve key
+        -> return (TPraosKeyEvolving, Left hk)
+        | otherwise
+        -> return (TPraosKeyAvailable hk, Right key)
+      TPraosNoKey ->
+        error "no KES key available"
+
+    -- | Evolve the given key so that its KES period matches @kesPeriod@.
+    evolveKey :: HotKey c -> HotKey c
+    evolveKey (HotKey oldPeriod outdatedKey) = go outdatedKey oldPeriod kesPeriod
+      where
+        go !sk c t
+          | t < c
+          = error "Asked to evolve KES key to old period"
+          | c == t
+          = HotKey kesPeriod sk
+          | otherwise
+          = case KES.updateKES () sk c of
+              Nothing  -> error "Could not update KES key"
+              Just sk' -> go sk' (c + 1) t
+
+    -- | PRECONDITION: we're in the 'TPraosKeyEvolving' node state.
+    saveNewKey :: HotKey c -> m ()
+    saveNewKey newKey = runUpdate updateForgeState $ \case
+      TPraosKeyEvolving -> Just (TPraosKeyAvailable newKey, ())
+      _                 -> error "must be in evolving state"
+
+{-------------------------------------------------------------------------------
+  Forging
+-------------------------------------------------------------------------------}
+
+forgeTPraosFields :: ( MonadRandom m
+                     , TPraosCrypto c
+                     , KES.Signable (KES c) toSign
+                     )
+                  => Update m (TPraosForgeState c)
+                  -> IsLeader (TPraos c)
+                  -> SL.KESPeriod
+                  -> (TPraosToSign c -> toSign)
+                  -> m (TPraosFields c toSign)
+forgeTPraosFields updateForgeState TPraosProof{..} kesPeriod mkToSign = do
+    hotKESKey <- evolveKESKeyIfNecessary updateForgeState (SL.KESPeriod kesEvolution)
```

Comment:
Do the key evolution in `forgeShelleyBlock` and move `forgeTPraosFields` back to the `Protocol` module.